Add settings to disable/enable MySQL/PostgreSQL module - Closes #4951

JavierJF opened the pull request:
# PR 4960: Comprehensive MySQL/PostgreSQL Module Control

## Main Purpose
This PR implements fine-grained control over ProxySQL's MySQL and PostgreSQL modules, allowing users to selectively enable or disable specific components via CLI arguments or configuration file settings. This addresses issue #4951 and provides better resource management and customization capabilities.

## New CLI Arguments Added
- **`--mysql-workers <true|false>`**: Enable/disable MySQL worker threads
- **`--pgsql-workers <true|false>`**: Enable/disable PostgreSQL worker threads  
- **`--mysql-admin <true|false>`**: Enable/disable MySQL admin interface
- **`--pgsql-admin <true|false>`**: Enable/disable PostgreSQL admin interface
- **`--mysql-monitor <true|false>`**: Enable/disable MySQL monitor module
- **`--pgsql-monitor <true|false>`**: Enable/disable PostgreSQL monitor module

## Configuration File Support
Corresponding boolean settings can be specified in the ProxySQL configuration file:
```ini
mysql-workers=true
mysql-admin=true  
mysql-monitor=true
pgsql-workers=true
pgsql-admin=true
pgsql-monitor=true
```

## Core Implementation Features

### 1. Selective Module Startup
- **Worker Threads**: Controls whether MySQL and/or PostgreSQL worker threads are created
- **Admin Interfaces**: Manages which admin interfaces (MySQL and/or PostgreSQL) listen for connections
- **Monitor Modules**: Controls whether MySQL and/or PostgreSQL monitoring threads are started

### 2. Intelligent Thread Management
- **MySQL Workers**: When disabled, no MySQL worker threads are created, reducing resource usage
- **PostgreSQL Workers**: When disabled, no PostgreSQL worker threads are created  
- **Admin Interfaces**: Each interface can be independently enabled/disabled
- **Monitor Modules**: Each monitor runs independently based on its setting

### 3. Admin Interface Variable Synchronization
- **Critical Fix**: Added synchronization between internal global variables and admin interface variables
- **`mysql-monitor_enabled`**: Now accurately reflects `--mysql-monitor` CLI argument state
- **`pgsql-monitor_enabled`**: Now accurately reflects `--pgsql-monitor` CLI argument state
- Ensures admin interface queries return correct module status

## Resource Management Benefits
- **Reduced Memory Usage**: Disabled modules don't allocate memory for threads and data structures
- **Lower CPU Usage**: No background processing for disabled modules  
- **Faster Startup**: Skip initialization of disabled components
- **Customizable Deployments**: Run only the modules you need

## Use Cases Enabled
1. **MySQL-Only Deployments**: Disable all PostgreSQL modules for pure MySQL environments
2. **PostgreSQL-Only Deployments**: Disable all MySQL modules for pure PostgreSQL environments  
3. **Admin-Only Mode**: Run admin interfaces without worker threads for configuration management
4. **Monitor-Only Mode**: Run monitoring without proxying capabilities
5. **Resource-Constrained Environments**: Disable unnecessary modules to save resources

## Comprehensive Testing Coverage

### Worker/Admin Module Tests (`reg_test_4960_modules_startup-t`)
- **16 test combinations** covering all possible enable/disable states
- **Port binding verification** ensures interfaces start/stop correctly
- **CLI argument and config file validation**
- Tests each combination: MySQL workers, PgSQL workers, MySQL admin, PgSQL admin

### Monitor Module Tests (`reg_test_4960_monitor_modules-t`)  
- **4 test combinations** for monitor module states
- **Database validation** via admin interface variables
- **Synchronization verification** between CLI args and runtime state

## Technical Implementation Details
- **Global Variables**: Added `mysql_workers`, `pgsql_workers`, `mysql_admin`, `pgsql_admin` to `ProxySQL_GlobalVariables`
- **Thread Control Logic**: Modified thread creation loops to respect module enable/disable flags
- **Admin Interface Management**: Updated interface initialization to check module flags
- **Configuration Processing**: Added CLI argument parsing and config file reading
- **Coherence Checks**: Added logging for module state consistency
- **Variable Synchronization**: Fixed admin interface to reflect actual module state

## Backward Compatibility
- All modules are **enabled by default**, maintaining existing behavior
- No breaking changes to existing configurations
- Graceful degradation when modules are disabled
- Clear error messages and logging for troubleshooting

## Documentation and Monitoring
- **Informative Logging**: Module status logged during startup
- **Admin Interface Visibility**: Module status queryable via `global_variables` table
- **Runtime Status**: Monitor module state accessible through admin interface

This implementation provides ProxySQL administrators with unprecedented control over which components run, enabling more efficient resource utilization and customized deployment scenarios while maintaining full backward compatibility.

JavierJF committed:
Remove non-supported checks from 'pgsql_replication_hostgroups' definition

JavierJF committed:
Add new variables for disable/enable MySQL/PostgreSQL modules - Closes #4951

Accessible as config file and command line switches:

  + --(mysql|pgsql)-workers=(true|false)
  + --(mysql|pgsql)-admin=(true|false)
  + --(mysql|pgsql)-monitor=(true|false)

The default value for all is 'true'.

JavierJF committed:
Remove old commented code

JavierJF committed:
Fix indentation and scope match for 'if' condition

Since editors sometimes don't completely disable non-reachable code,
like macro-protected, it's cleaner too keep parenthesis open/close
outside this. Otherwise this can confuse editors scope matching.

JavierJF committed:
Fix 'MySQLWorker' threads naming

renecannao commented:
Can one of the admins verify this patch?

mirostauder commented:
retest this please

renecannao commented on include/ProxySQL_Admin_Tables_Definitions.h:
```diff
@@ -276,7 +276,7 @@
 #define ADMIN_SQLITE_TABLE_PGSQL_QUERY_RULES_FAST_ROUTING  "CREATE TABLE pgsql_query_rules_fast_routing (username VARCHAR NOT NULL , database VARCHAR NOT NULL , flagIN INT NOT NULL DEFAULT 0 , destination_hostgroup INT CHECK (destination_hostgroup >= 0) NOT NULL , comment VARCHAR NOT NULL , PRIMARY KEY (username, database, flagIN) )"
 #define ADMIN_SQLITE_TABLE_PGSQL_HOSTGROUP_ATTRIBUTES "CREATE TABLE pgsql_hostgroup_attributes (hostgroup_id INT NOT NULL PRIMARY KEY , max_num_online_servers INT CHECK (max_num_online_servers>=0 AND max_num_online_servers <= 1000000) NOT NULL DEFAULT 1000000 , autocommit INT CHECK (autocommit IN (-1, 0, 1)) NOT NULL DEFAULT -1 , free_connections_pct INT CHECK (free_connections_pct >= 0 AND free_ ...
-#define ADMIN_SQLITE_TABLE_PGSQL_REPLICATION_HOSTGROUPS "CREATE TABLE pgsql_replication_hostgroups (writer_hostgroup INT CHECK (writer_hostgroup>=0) NOT NULL PRIMARY KEY , reader_hostgroup INT NOT NULL CHECK (reader_hostgroup<>writer_hostgroup AND reader_hostgroup>=0) , check_type VARCHAR CHECK (LOWER(check_type) IN ('read_only','innodb_read_only','super_read_only','read_only|innodb_read_only','r ...
+#define ADMIN_SQLITE_TABLE_PGSQL_REPLICATION_HOSTGROUPS "CREATE TABLE pgsql_replication_hostgroups (writer_hostgroup INT CHECK (writer_hostgroup>=0) NOT NULL PRIMARY KEY , reader_hostgroup INT NOT NULL CHECK (reader_hostgroup<>writer_hostgroup AND reader_hostgroup>=0) , check_type VARCHAR CHECK (LOWER(check_type) IN ('read_only')) NOT NULL DEFAULT 'read_only' , comment VARCHAR NOT NULL DEFAULT '' ...
```
Since 3.0.1 we can't simply change table definition. We need an upgrade procedure.

renecannao replied:
Actually, this may need a PR on its own.

JavierJF replied:
I don't know if that is really worth it, the check type is never used in the PostgreSQL implementation, since only `read_only` (the default) is supported, which value is never really used either. A migration procedure for a unused settings seems unnecessary... But I can split it into another PR.

renecannao replied:
It is necessary because during bootstrap ProxySQL will delete any table that do not have the expected table definition. So even removing unused constraint should cause ProxySQL to drop and recreate the table.

JavierJF replied:
Ahhh right! now I remember. The table is being rebuild automatically, so the entries already stored will be missing without the upgrade due to the rebuilding. Should I pack those changes in another PR?

renecannao replied:
Yes, a dedicated PR is better, as it has nothing to do with start-up options.
Thanks!

renecannao commented on include/proxysql_glovars.hpp:
```diff
@@ -88,6 +88,10 @@ class ProxySQL_GlobalVariables {
+		bool mysql_workers;
+		bool pgsql_workers;
+		bool mysql_admin;
+		bool pgsql_admin;
```
nitpicking: we should follow a pattern, either first all mysql variables then all posgres variables (or viceversa), or follow some order: `*_monitor` , `*-workers` , `*_admin` .

renecannao commented:
When documenting these variables it is worth to mention that the Hostgroups Manager and Authentication Module are always enabled, no matter if Admin/Monitor/Workers are disabled.
As an example, either Monitor or Workers can access Hostgroups Manager even if the other module is disabled.

JavierJF committed:
Revert "Remove non-supported checks from 'pgsql_replication_hostgroups' definition"

This reverts commit 80581625c851027a7dc0612f500858066b47da71. Since the
table changes require additional code due to the standard table
migration procedure, these changes shall be move to their own PR.

JavierJF committed:
Improve variable definition order in 'ProxySQL_GlobalVariables::global'

JavierJF committed:
Add initial doc for variables to disable/enable MySQL/PostgreSQL modules

Should be complemented with more details in the official doc.

renecannao committed:
Add comprehensive TAP test for module enable/disable functionality

- Tests all 16 combinations of MySQL/PgSQL worker and admin module configurations
- Verifies both admin and worker port binding behavior
- Includes aggressive cleanup to prevent interference between test cases
- Tests both CLI arguments and config file settings
- Uses dedicated port range (13750-13813) to avoid conflicts
- Includes prerequisite checking for required system tools

renecannao committed:
Fix test logic errors and compiler warnings

- Correct test case expectations to match CLI arguments behavior
- Fix binary naming vs actual module enable/disable logic
- Ensure admin and worker modules are tested independently
- Resolve unused result warnings from system() calls
- Remove debug logging to clean up test output
- Validate that all module combinations work correctly

renecannao committed:
Optimize test performance by removing unnecessary sleep

- Remove redundant sleep(1) between port check attempts
- Port checking with netcat completes instantly
- Reduces test execution time by approximately 60+ seconds
- Maintains same functionality while improving performance

renecannao committed:
Add monitor module enable/disable testing

Add comprehensive TAP test for verifying MySQL and PostgreSQL monitor
modules can be enabled/disabled via CLI arguments. Test validates monitor
status by querying global_variables table for mysql-monitor_enabled and
pgsql-monitor_enabled variables.

renecannao committed:
Redirect ProxySQL output to log files in test datadirs

Modify monitor test to redirect ProxySQL stdout and stderr to
proxysql.log files in each test's datadir for cleaner console output
and better debugging capabilities.

renecannao committed:
Fix monitor module synchronization with admin interface variables

Add synchronization code to ensure mysql-monitor_enabled and pgsql-monitor_enabled
variables in global_variables table accurately reflect the state of CLI arguments
--mysql-monitor and --pgsql-monitor. Previously, CLI arguments correctly disabled
monitor modules but admin interface variables remained true, causing inconsistent
state and test failures.

The fix adds synchronization after admin database initialization to update
admin interface variables based on internal global variables, ensuring proper
reflection of monitor module state.

renecannao committed:
Add module enable/disable TAP tests to default test group

Add reg_test_4960_modules_startup-t and reg_test_4960_monitor_modules-t to the
default-g1 test group to ensure comprehensive testing of the module enable/disable
functionality. Tests run only in default group to avoid duplication across multiple
test groups.

renecannao committed:
Merge branch 'v3.0' into v3.0-4951

Signed-off-by: René Cannaò <rene@proxysql.com>

renecannao commented:
Enhanced PR description with comprehensive feature overview including CLI arguments, configuration options, use cases, testing coverage, and implementation details.

renecannao commented:
## ⚠️ Important Dependency Notice

This PR has potential dependencies with **PR 5217** (Fix issue 5186: Prevent admin query crashes after PROXYSQL STOP/START cycles). Both PRs should be tested together before merge.

### Overlap Areas:
1. **lib/ProxySQL_Admin.cpp** - Both PRs modify this file
   - This PR: Fixes admin interface variable synchronization
   - PR 5217: Adds null pointer protection in `save_*_from_runtime()` functions

2. **Module Lifecycle Management**
   - This PR: Allows selective module disable/enable at startup
   - PR 5217: Implements STOP/START cycles assuming all modules are present

3. **src/main.cpp** - Both PRs modify startup behavior

4. **Global Variable Handling**
   - This PR: Adds module-specific flags (`mysql_workers`, `pgsql_workers`, etc.)
   - PR 5217: Uses `GloVars.global.nostart` for restart sequences

### Verification Needed:
- [ ] Test all module enable/disable combinations with PROXYSQL STOP/START
- [ ] Verify null pointer protection works with disabled modules
- [ ] Check admin interface variable accuracy in mixed states
- [ ] Validate no crashes during STOP/START with selective modules

**Tracking Issue**: #5218 has been created to coordinate this verification.

Please do not merge until dependencies are verified.
